src/grpc/infer_handler.h

```diff
@@ -652,12 +651,143 @@ class InferHandlerState {
       ctx_->set_compression_level(compression_level);
     }
 
+    void GrpcContextAsyncNotifyWhenDone(InferHandlerStateType* state)
+    {
+      ctx_->AsyncNotifyWhenDone(state);
+    }
+
+    bool IsCancelled() { return ctx_->IsCancelled(); }
+
     // Increments the ongoing request counter
     void IncrementRequestCounter() { ongoing_requests_++; }
 
     // Decrements the ongoing request counter
     void DecrementRequestCounter() { ongoing_requests_--; }
 
+
+    // Inserts the state to a set tracking active requests
+    // within the server core.
+    void InsertInflightState(
+        InferHandlerStateType* state, TRITONSERVER_InferenceRequest* irequest)
+    {
+      std::lock_guard<std::mutex> lock(mu_);
+      state->irequest_ptr_ = irequest;
+      inflight_states_.insert(state);
+    }
+
+    // Erases the state to a set tracking active requests
+    // within the server core.
+    void EraseInflightState(InferHandlerStateType* state)
+    {
+      std::lock_guard<std::mutex> lock(mu_);
+      inflight_states_.erase(state);
+    }
+
+    // Issues the cancellation for all inflight requests
+    // being tracked by this context.
+    TRITONSERVER_Error* IssueRequestCancellation()
+    {
+      TRITONSERVER_Error* err = nullptr;
+      {
+        std::lock_guard<std::mutex> lock(mu_);
+
+        // Issues the request cancellation to the core.
+        for (auto state : inflight_states_) {
+          LOG_VERBOSE(1) << "Issuing cancellation for " << state->unique_id_;
+          if (state->irequest_ptr_ == nullptr) {
+            continue;
+          }
+          bool is_cancelled = false;
+          err = TRITONSERVER_InferenceRequestIsCancelled(
+              state->irequest_ptr_, &is_cancelled);
+          if ((err == nullptr) && (!is_cancelled)) {
+            err = TRITONSERVER_InferenceRequestCancel(state->irequest_ptr_);
+          }
+          if (err != nullptr) {
+            const char* request_id = "";
+            TRITONSERVER_Error* tmp_err = TRITONSERVER_InferenceRequestId(
+                state->irequest_ptr_, &request_id);
+            if (tmp_err != nullptr) {
+              // Ignore if failed to get request id.
+              TRITONSERVER_ErrorDelete(tmp_err);
+            }
+            TRITONSERVER_Error* wrapped_err = TRITONSERVER_ErrorNew(
+                TRITONSERVER_ERROR_INTERNAL,
+                std::string(
+                    std::string("[request id:") + request_id +
+                    "]: " + TRITONSERVER_ErrorMessage(err))
+                    .c_str());
+            TRITONSERVER_ErrorDelete(err);
```

Comment:
nit: can we make this a helper function (e.g., `AppendRequestIdToError`)?

Reply:
I have some more ideas to improve the code. I propose we improve it in the next iteration.